### DISCUSSION TITLE
How add specifc-accessor for new accelerator on SYCL？

### DISCUSSION THREAD

**Discussion by wangzy0327** [intel/llvm#16000] [Q&A]
How to add accelerator hardware to SYCL, where new device-specific RAM support is required for kernel operations?

Are there any documents or developer manuals that I can refer to?

@JackAKirk

**Comment by JackAKirk** [comment]
I don't know. Maybe check out https://github.com/codeplaysoftware/oneapi-construction-kit

**Comment by AlexeySachkov** [comment]
> How to add accelerator hardware to SYCL, where new device-specific RAM support is required for kernel operations?

I think that it heavily depends on what exactly needs to be done.

If you need to generate some special ASM code for those operations like `operator[]`, then you need to start working upwards, i.e. what are the inputs for your device compiler (LLVM IR, SPIR-V, something else) and see what you need to generate there - it could be a new LLVM pass, or it could be just a modification to SYCL headers.

If you need to emit kernel arguments differently, then I suggest you to look into how `accessor` is implemented: it is annotated with a special attribute to be able to recognize it in FE and emit the right number of kernel arguments of the right type for it, plus some integration header bits. Once that's done, you can detect at RT that new type of accessor is used as a kernel argument and call right runtime APIs to set the right arguments or do any special-handling you need.
`work_group_memory` extension implemented in #15178 is essentially a special case of local accessor and you can use this PR as an example here.
